lib/src/main/java/graphql/nadel/engine/transform/partition/NadelPartitionTransform.kt

```diff
@@ -0,0 +1,266 @@
+        val pathToSplitPoint = routingDirective.getArgument("pathToSplitPoint")
+            ?.argumentValue
+            ?.value as ArrayValue?
+
+        return pathToSplitPoint?.values?.map { it as StringValue }?.map { it.value }
+    }
+
+    override suspend fun transformField(
+        executionContext: NadelExecutionContext,
+        serviceExecutionContext: NadelServiceExecutionContext,
+        transformer: NadelQueryTransformer,
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        service: Service,
+        field: ExecutableNormalizedField,
+        state: State,
+    ): NadelTransformFieldResult {
+
+        if (state.errors != null) {
+            // At this point we know we can't partition the field, but removing it could result
+            // in a query with no top-level fields, which would be invalid.
+            // Adding __typename as an artificial field also wouldn't work because that field
+            // belongs to all operation types, which causes issues further down the execution.
+            return NadelTransformFieldResult.unmodified(field)
+        }
+
+        val fieldPartitions = checkNotNull(state.fieldPartitions) { "Expected fieldPartitions to be set" }
+
+        val primaryPartition = fieldPartitions.values.first()
+
+        // TODO: throw error if operation is Subscription?
+        val rootType = executionContext.query.operation.getType(executionBlueprint.engineSchema)
+
+        val partitionCalls = coroutineScope {
+            fieldPartitions.values.drop(1).map {
+                async {
+                    val topLevelField = NFUtil.createField(
+                        executionBlueprint.engineSchema,
+                        rootType,
+                        // TODO: queryPath contains field aliases, not field names, which results in an error.
+                        field.queryPath,
+                        it.normalizedArguments,
+                        it.children
+                    )
+
+                    engine.executePartitionedCall(topLevelField, service, state.executionContext)
+                }
+            }
+        }
+
+        state.partitionCalls.addAll(partitionCalls)
+
+        return NadelTransformFieldResult(newField = primaryPartition)
+    }
+
+    override suspend fun getResultInstructions(
+        executionContext: NadelExecutionContext,
+        serviceExecutionContext: NadelServiceExecutionContext,
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        service: Service,
+        overallField: ExecutableNormalizedField,
+        underlyingParentField: ExecutableNormalizedField?,
+        result: ServiceExecutionResult,
+        state: State,
+        nodes: JsonNodes,
+    ): List<NadelResultInstruction> {
+        val parentNodes = nodes.getNodesAt(
+            queryPath = underlyingParentField?.queryPath ?: NadelQueryPath.root,
+            flatten = true,
+        )
```

Comment:
Maybe a `.singleOrNull() ?: return emptyList()`

So we do things like `parentNodes.first()` later